[#5] 로그인 API 및 jwt 토큰 발급 및 검증 기능 추가

rbsks opened the pull request:
### 추가사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
- 로그인 API 추가
- jwt 토큰 발급 및 검증 추가

### 테스트
- [x] 테스트 코드
- [x] API 테스트

rbsks committed:
Feat: 로그인 API 및 jwt 토큰 발급 및 검증 기능 추가

f-lab-kyle commented on src/main/java/com/stylelab/common/security/WebSecurityConfiguration.java:
```diff
@@ -0,0 +1,114 @@
+                .authorizeRequests()
+                .requestMatchers(CorsUtils::isPreFlightRequest).permitAll()
+                .antMatchers(
+                        "/v1/users/signup",
+                        "/v1/users/check-email",
+                        "/v1/users/check-nickname",
+                        "/v1/users/signin"
```
v2같은 버저닝, 그리고 users 뒤에 오는 path에 대해 전부 처리를 해야 할 상황이라면 `*`을 사용하셔도 좋을것 같습니다.

rbsks replied:
users api를 전부 permit all로 설정할 게 아니라서 버저닝 부분만 멘토님 말씀대로 처리하도록 하겠습니다!

f-lab-kyle commented on src/main/java/com/stylelab/common/security/WebSecurityConfiguration.java:
```diff
@@ -0,0 +1,114 @@
+                    response.setStatus(HttpStatus.FORBIDDEN.value());
+                    response.setCharacterEncoding(StandardCharsets.UTF_8.name());
+                    response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+                    response.getWriter().write("{"
+                            + "\"code\": \"" + FORBIDDEN.getCode() + "\","
+                            + "\"message\": \""  + FORBIDDEN.getMessage() + "\""
+                            +"}"
+                    );
```
가급적 코드로 생성하는게 좋지 않을까 생각되네요 (json 파싱 오류, 응답 객체 변경 등)

rbsks replied:
넵 리팩토링하도록 하겠습니다 ㅎㅎ

f-lab-kyle commented on src/main/java/com/stylelab/common/security/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,46 @@
+        filterChain.doFilter(request, response);
+    }
+
+    public String resolveTokenFromRequest(HttpServletRequest request) {
```
public일 이유가 있으려나요, 그리고 null을 리턴하는 메소드에 대해서는 항상 optional을 고민하시는게 좋을것 같아요 (언어적 특성)

rbsks replied:
Optional로 반환하는게 if 문도 제거할 수 있고 좋겠네요 ㅎㅎ

f-lab-kyle commented on src/main/java/com/stylelab/common/security/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,92 @@
+
+import com.stylelab.common.exception.ServiceException;
+import com.stylelab.common.security.service.CustomUserDetailsService;
+import io.jsonwebtoken.*;
```
https://dev.to/kingori/import-5531
https://invicr.github.io/2020/03/03/IntelliJ-Import-%EC%84%A4%EC%A0%95/

f-lab-kyle commented on src/main/java/com/stylelab/user/application/UsersFacade.java:
```diff
@@ -19,22 +28,34 @@ public class UsersFacade {
-    public void signup(final SignupRequestDto signupRequestDto) {
-        if (!Objects.equals(signupRequestDto.getPassword(), signupRequestDto.getConfirmPassword())) {
+    public void signup(final SignupRequest signupRequest) {
+        if (!Objects.equals(signupRequest.getPassword(), signupRequest.getConfirmPassword())) {
```
StringUtils.equals 를 사용하는게 좋을까요 Objects.equals를 사용하는게 좋을까요?

rbsks replied:
이 부분은 공부해 보고 블로깅 하도록 하겠습니다! 감사합니다 ㅎㅎ

rbsks replied:
혹시 Apache commons Lang의 StringUtils를 말씀하시는 건가요??!

f-lab-kyle commented on src/main/java/com/stylelab/user/repository/UsersRepository.java:
```diff
@@ -3,8 +3,11 @@
 
     boolean existsByEmail(String email);
     boolean existsByNickname(String nickname);
+    Optional<Users> findByEmail(String email);
```
위 existsByEmail 와 실행 메커니즘이 뭐가 다를까요? (findByEmail로 exists처리를 할 수 있지 않을까 하여)

rbsks replied:
findByEmail은 User table의 컬럼을 전부 다 가져오는 반면에 existsByEmail은 userId 한 개만 가져와서 존재 여부를 판별하기 때문에 더 효율적이고 두 개의 쿼리의 역할이 다르다 판단하여 분리하였습니다.

f-lab-kyle commented on src/test/java/com/stylelab/common/security/jwt/JwtTokenProviderTest.java:
```diff
@@ -0,0 +1,73 @@
+    private JwtTokenProvider jwtTokenProvider;
+
+    @Test
+    @DisplayName("토큰 생성 테스트")
```
테스트의 이름은 문장의 형태가 좋을것 같아요. 상황에 따른 결과값과 기대값에 대한 비교.

f-lab-kyle reviewed (commented):
리뷰 받고싶은 항목이 별도로 있다면 셀프 댓글로 달아주셔도 좋습니다. 이 부분은 멘토링 시간때 이야기 나눠봐요.

rbsks committed:
Refactor: 코드 리뷰 반영
- security accessDeniedHandler, authenticationEntryPoint 추가
- security antMatchers 버저닝 와일드 카드로 수정
- intellij import 설정 추가
- CI/CD job name 수정
